### PR TITLE
15.0 fix leads activity filter pmo

### PR DESCRIPTION
**Pull request by madprog**
**[FIX] mail: consider more than last 1000 activities**

Because `activities_to_check` was assigned in each loop, the result was only containing the last chunk of 1000 records, instead of the whole result.

This commit fixes the issue by concatenating into the list instead of just assigning it.

**[FIX] mail: fix performance issue on activities filter**

Accessing `valid_docs.ids` for each activity to check was costly, and in my example the whole loop took 125 seconds to execute.

By turning it into a set, the same loop takes 0.2 seconds to execute.

When re-constructing the list based on ids to keep the order, using a set is also faster, and takes another 0.2 seconds instead of 13.6.

### CODE REVIEW

File: addons/mail/models/mail_activity.py [hunk 1]
```diff
@@ -356,7 +356,8 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
             self._cr.execute("""
                 SELECT DISTINCT activity.id, activity.res_model, activity.res_id
                 FROM "%s" activity
-                WHERE activity.id = ANY (%%(ids)s)""" % self._table, dict(ids=list(sub_ids)))
+                WHERE activity.id = ANY (%%(ids)s)
+                AND activity.res_id != 0""" % self._table, dict(ids=list(sub_ids)))
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_activity.py:360]: How could this happen ? This is a required field, and always used through a document owning the activity.

File: addons/mail/models/mail_activity.py [hunk 2]
```diff
@@ -376,14 +376,16 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
             right = DocumentModel.check_access_rights(doc_operation, raise_exception=False)
             if right:
                 valid_docs = DocumentModel.browse(doc_ids)._filter_access_rules(doc_operation)
+                valid_doc_ids = set(valid_docs.ids)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_activity.py:379]: Note that it could be inlined directly below

File: addons/mail/models/mail_activity.py [hunk 3]
```diff
@@ -376,14 +376,16 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
...
                 allowed_ids += [
                     activity['id'] for activity in activities_to_check
-                    if activity['res_model'] == doc_model and activity['res_id'] in valid_docs.ids]
+                    if activity['res_model'] == doc_model and activity['res_id'] in valid_doc_ids]
 
         if count:
             return len(allowed_ids)
         else:
             # re-construct a list based on ids, because 'allowed_ids' does not keep the original order
+            allowed_ids = set(allowed_ids)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_activity.py:388]: Note that it could be inlined directly below

**Review by awa-odoo — COMMENTED** [review]
Hello @madprog 

Looks good to me except the "res_id != 0" part.
Are there really records in mail.activity where res_id is 0?
Shouldn't those be cleaned? Not sure what they're about.

Thanks for fixing it!

cc @tde-banana-odoo

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the PR, had a first look :)

Side note: isn't it valid for previous stables versions ?

File: addons/mail/models/mail_activity.py [hunk 4]
```diff
@@ -375,9 +375,10 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
             right = DocumentModel.check_access_rights(doc_operation, raise_exception=False)
             if right:
                 valid_docs = DocumentModel.browse(doc_ids)._filter_access_rules(doc_operation)
-                allowed_ids += [
+                valid_doc_ids = set(valid_docs.ids)
+                allowed_ids.update({
```
Comment (KangOl) [inline addons/mail/models/mail_activity.py:379]: Directly use the generator
```suggestion
                allowed_ids.update(
```

File: addons/mail/models/mail_activity.py [hunk 5]
```diff
@@ -375,9 +375,10 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
...
                     activity['id'] for activity in activities_to_check
-                    if activity['res_model'] == doc_model and activity['res_id'] in valid_docs.ids]
+                    if activity['res_model'] == doc_model and activity['res_id'] in valid_doc_ids})
```
Comment (KangOl) [inline addons/mail/models/mail_activity.py:381]:
```suggestion
                    if activity['res_model'] == doc_model and activity['res_id'] in valid_doc_ids)
```

File: addons/mail/models/mail_activity.py (hunk 2 above)
Comment (madprog) [reply addons/mail/models/mail_activity.py:378]: By inlining it in the generator expression, won't the set be created at each loop?

**Review by KangOl — CHANGES_REQUESTED** [review]

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**